core/src/main/scala/com/softwaremill/react/kafka_api_opt3/Consumer.scala

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright (C) 2014 - 2016 Softwaremill <http://softwaremill.com>
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package com.softwaremill.react.kafka_api_opt3
+
+import scala.collection.immutable
+import scala.concurrent.duration.FiniteDuration
+import akka.stream.Inlet
+import org.apache.kafka.common.TopicPartition
+import akka.stream.stage.GraphStageWithMaterializedValue
+import org.apache.kafka.clients.consumer.OffsetAndMetadata
+import akka.stream.Outlet
+import org.apache.kafka.clients.consumer.ConsumerRecord
+import akka.stream.SourceShape
+import org.apache.kafka.clients.consumer.KafkaConsumer
+import akka.stream.Shape
+import akka.stream.scaladsl.GraphDSL
+import akka.stream.scaladsl.Source
+import akka.stream.scaladsl.Flow
+import scala.concurrent.Future
+import akka.NotUsed
+import akka.Done
+
+object Consumer {
+
+  trait Message[K, V] {
+    def transaction: Transaction
+    def record: ConsumerRecord[K, V]
+  }
+
+  trait Transaction {
+    def commit(): Future[Done]
```

Comment:
I don't see otherwise working in face of not-strictly ordered acknowledgements. I.e. if someone commits M2 but not yet M1 then it would be just as surprising to see M1 committed in the end. I don't know maybe we can make this configurable?

As for move away from Kafka, we should do it sparingly, but in the end we would like to see an Acknowledgement API that can go into akka-streams and provide an interop layer between various projects that can handle these acknowledged operations. Anyway, we have two options for the semantics of out-of-order commits sketched here already :)